### PR TITLE
Add image GPU loading in preprocimage

### PR DESCRIPTION
**Pull request by ClaroHenrique**
This PR aims to optimize the GPU algorithm by loading the TI and AUXTI image only once in `preprocimage`.

### CODE REVIEW

File: src/imfilter.jl [hunk 1]
```diff
@@ -7,6 +7,10 @@ function imfilter_cpu(img, krn)
 end
 
 function imfilter_gpu(img, krn)
+  imfilter_gpu(img |> CuArray, krn)
+end
+
+function imfilter_gpu(img::CuArray, krn)
```
Comment (juliohm) [inline src/imfilter.jl:13]:
```suggestion
```

File: src/utils.jl [hunk 2]
```diff
@@ -54,11 +54,32 @@ function activation(hard, tile)
...
+end
+
+const view_kernel = CUDA.functional() ? view_gpu : view_cpu
+
 function imagepreproc(trainimg, soft, geoconfig)
   padsize = geoconfig.padsize
 
   TI = Float64.(trainimg)
   replace!(TI, NaN => 0.)
+  TI_kernel = TI |> array_kernel
```
Comment (juliohm) [inline src/utils.jl:82]:
```suggestion
  TI_kernel = array_kernel(TI)
```

File: src/utils.jl [hunk 3]
```diff
@@ -54,11 +54,32 @@ function activation(hard, tile)
...
+
+const array_kernel = CUDA.functional() ? array_gpu : array_cpu
+
+function view_cpu(array, I)
+  view(array, I)
+end
+
+function view_gpu(array, I)
+  array[I] |> Array
+end
```
Comment (juliohm) [inline src/utils.jl:73]:
```suggestion
array_cpu(array) = array

array_gpu(array) = CuArray(array)

const array_kernel = CUDA.functional() ? array_gpu : array_cpu

view_cpu(array, I) = view(array, I)

view_gpu(array, I) = Array(array[I])
```

File: src/utils.jl [hunk 4]
```diff
@@ -70,10 +91,12 @@ function imagepreproc(trainimg, soft, geoconfig)
     replace!(AUX, NaN => 0.)
     replace!(AUXTI, NaN => 0.)
 
-    AUX, AUXTI
+    AUXTI_kernel = AUXTI |> array_kernel
+
+    AUX, AUXTI_kernel
```
Comment (juliohm) [inline src/utils.jl:96]:
```suggestion
    AUX, array_kernel(AUXTI)
```

**Review by juliohm — CHANGES_REQUESTED** [review]

File: src/imfilter.jl [hunk 5]
```diff
@@ -16,7 +16,7 @@ function imfilter_gpu(img, krn)
   padkrn  = padarray(krn, Fill(zero(T), ntuple(i->0, N), padsize))
 
   # perform ifft(fft(img) .* conj.(fft(krn)))
-  fftimg = img |> CuArray |> CUFFT.fft
+  fftimg = img |> CUFFT.fft
   fftkrn = padkrn |> CuArray |> CUFFT.fft
```
Comment (juliohm) [inline src/imfilter.jl:20]: I thought that we would also move the kernel only once to the GPU as a buffer with the appropriate padding already. Did I misunderstood the idea?

File: src/utils.jl [hunk 6]
```diff
@@ -70,10 +83,10 @@ function imagepreproc(trainimg, soft, geoconfig)
     replace!(AUX, NaN => 0.)
     replace!(AUXTI, NaN => 0.)
 
-    AUX, AUXTI
+    AUX, array_kernel(AUXTI)
```
Comment (juliohm) [inline src/utils.jl:86]: Why AUX is not being moved as well? Can we move all arrays at the very end of the function instead? So keep the code untouched and only at the end before returning you can call the array_kernel in everything.

**Comment by juliohm** [comment]
Left a few comments @ClaroHenrique 👍🏽

File: src/imfilter.jl (hunk 5 above)
Comment (ClaroHenrique) [reply src/imfilter.jl:20]: I thought this kernel optimization would be implemented in the next PR.

File: src/utils.jl (hunk 6 above)
Comment (ClaroHenrique) [reply src/utils.jl:86]: The AUX is not directly used in the `imfilter`, instead, only a tile is used as kernel. So its not worth the conversion.

File: src/imfilter.jl (hunk 5 above)
Comment (juliohm) [reply src/imfilter.jl:20]: Ok, let's do it in a separate PR then. I thought that this PR was about loading all arrays in the GPU, but we can do it in the next PR too.

File: src/utils.jl (hunk 6 above)
Comment (juliohm) [reply src/utils.jl:86]: Makes sense. I think it is worth calling `array_kernel` at the end of the function anyways to make the code easier to follow and modify in the future.

Comment (juliohm) [reply src/utils.jl:86]: At the end you could simply return something like:

```julia
# move arrays to GPU if necessary
TIₖ = array_kernel(TI)
SOFTₖ = zip(first.(SOFT), array_kernel.(last.(SOFT)))

TIₖ, SOFTₖ

**Pull request merged**